packages/multichain/src/caip25Permission.ts

```diff
@@ -0,0 +1,283 @@
+  validator: PermissionValidatorConstraint;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+type Caip25EndowmentSpecificationBuilderOptions = {
+  methodHooks: {
+    findNetworkClientIdByChainId: (chainId: Hex) => NetworkClientId;
+    listAccounts: () => { address: Hex }[];
+  };
+};
+
+/**
+ * Helper that returns a `endowment:caip25` specification that
+ * can be passed into the PermissionController constructor.
+ *
+ * @param builderOptions - The specification builder options.
+ * @param builderOptions.methodHooks - The RPC method hooks needed by the method implementation.
+ * @returns The specification for the `caip25` endowment.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.Endowment,
+  Caip25EndowmentSpecificationBuilderOptions,
+  Caip25EndowmentSpecification
+> = ({ methodHooks }: Caip25EndowmentSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.Endowment,
+    targetName: Caip25EndowmentPermissionName,
+    allowedCaveats: [Caip25CaveatType],
+    endowmentGetter: (_getterOptions?: EndowmentGetterParams) => null,
+    validator: (permission: PermissionConstraint) => {
+      const caip25Caveat = permission.caveats?.[0];
+      if (
+        permission.caveats?.length !== 1 ||
+        caip25Caveat?.type !== Caip25CaveatType
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Invalid caveats. There must be a single caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+
+      const { requiredScopes, optionalScopes, isMultichainOrigin } =
+        caip25Caveat.value as Caip25CaveatValue;
+
+      if (
+        !requiredScopes ||
+        !optionalScopes ||
+        typeof isMultichainOrigin !== 'boolean'
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Received invalid value for caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+
+      const { normalizedRequiredScopes, normalizedOptionalScopes } =
+        validateAndNormalizeScopes(requiredScopes, optionalScopes);
+
+      const isChainIdSupported = (chainId: Hex) => {
+        try {
+          methodHooks.findNetworkClientIdByChainId(chainId);
+          return true;
+        } catch (err) {
+          return false;
+        }
+      };
+
+      assertScopesSupported(normalizedRequiredScopes, {
+        isChainIdSupported,
+      });
+      assertScopesSupported(normalizedOptionalScopes, {
+        isChainIdSupported,
+      });
+
+      // Fetch EVM accounts from native wallet keyring
+      // These addresses are lowercased already
+      const existingEvmAddresses = methodHooks
+        .listAccounts()
+        .map((account) => account.address);
+      const ethAccounts = getEthAccounts({
+        requiredScopes: normalizedRequiredScopes,
+        optionalScopes: normalizedOptionalScopes,
+      }).map((address) => address.toLowerCase() as Hex);
```

Comment:
How do we know this is `Hex`? Should we update `getEthAccounts` to verify this?

Reply:
Addressed here: https://github.com/MetaMask/core/pull/4784/commits/a53240985b8abddf7a75cc183bd918b8e4d9f1ec